src/Xamarin.Android.Build.Tasks/Tasks/ResolveAssemblies.cs

```diff
@@ -223,22 +223,28 @@ void AddAssemblyReferences (MetadataResolver resolver, Dictionary<string, ITaskI
 			indent += 2;
 
 			// Add this assembly
+			ITaskItem assemblyItem = null;
 			if (topLevel) {
-				if (!string.IsNullOrEmpty (targetFrameworkIdentifier) && assemblies.TryGetValue (assemblyName, out ITaskItem taskItem)) {
-					if (string.IsNullOrEmpty (taskItem.GetMetadata ("TargetFrameworkIdentifier"))) {
-						taskItem.SetMetadata ("TargetFrameworkIdentifier", targetFrameworkIdentifier);
+				if (assemblies.TryGetValue (assemblyName, out assemblyItem)) {
+					if (!string.IsNullOrEmpty (targetFrameworkIdentifier) && string.IsNullOrEmpty (assemblyItem.GetMetadata ("TargetFrameworkIdentifier"))) {
+						assemblyItem.SetMetadata ("TargetFrameworkIdentifier", targetFrameworkIdentifier);
 					}
 				}
 			} else {
-				assemblies [assemblyName] = CreateAssemblyTaskItem (assemblyPath, targetFrameworkIdentifier);
+				assemblies [assemblyName] = 
+					assemblyItem = CreateAssemblyTaskItem (assemblyPath, targetFrameworkIdentifier);
 			}
 
 			// Recurse into each referenced assembly
 			foreach (var handle in reader.AssemblyReferences) {
 				var reference = reader.GetAssemblyReference (handle);
 				string reference_assembly;
 				try {
-					reference_assembly = resolver.Resolve (reader.GetString (reference.Name));
+					var referenceName = reader.GetString (reference.Name);
+					if (assemblyItem != null && referenceName == "Mono.Android") {
+						assemblyItem.SetMetadata ("MonoAndroidReference", "True");
```

Comment:
I fear I might start heading toward [“Hungarian”](https://en.m.wikipedia.org/wiki/Hungarian_notation) with this suggestion, but should this be `%(HasMonoAndroidReference)`, so that *in context* we know that the value is supposed to be Boolean?